File: mabi/media/damage/serializer.js
mabi.Serializer = function(){
};

mabi.Serializer.prototype.serialize = function(object){
    var this_ = this;
    if(object instanceof mabi.Effect){
        return this.simple(object, ['op', 'param', 'min']);
    }else if(object instanceof mabi.Element){
        return this.serializeElement(object);
    }else if(object instanceof Array){
        var result = [];
        $.each(object, function(i, v){
            result.push(this_.serialize(v));
        });
        return result;
    }
    throw '未対応';
};

mabi.Serializer.prototype.deserializeEffect = function(json){
    var this_ = this;
    return new mabi.Effect(json);
};

mabi.Serializer.prototype.deserializeElement = function(json){
    var this_ = this;
    var result = new (this.type(json.type))({
        name: json.name
    });

    if(json.effects){
        $.each(json.effects, function(i, effect){
            result.addEffect(this_.deserializeEffect(effect));
        });
    }
    
    if(json.children){
        $.each(json.children, function(i, child){
            result.addChild(
                this_.deserializeElement(child),
                child.slot
            );
        });
    }
    return result;
};


// ----------------------------------------------------------------------
// private

(function(){
    var map = {
        Mob: mabi.Mob,
        Element: mabi.Element,
        EquipmentSet: mabi.EquipmentSet,
        Equipment: mabi.Equipment
    };

    mabi.Serializer.NAME_TO_TYPE = map;
})();

/**
 * Element の型名を取得する
 */
mabi.Serializer.prototype.typeName = function(element){
    var result;
    $.each(mabi.Serializer.NAME_TO_TYPE, function(name, type){
        if(element.constructor == type){
            result = name;
            return false;
        }
        return true;
    });
    if(!result) throw '未対応の Element タイプです';
    return result;
};

/**
 * Element の型を取得する
 */
mabi.Serializer.prototype.type = function(name){
    var type = mabi.Serializer.NAME_TO_TYPE[name];
    if(!type) throw '未対応の型です:' + name;
    return type;
};

/**
 * 
 */
mabi.Serializer.prototype.serializeElement = function(object){
    var this_  = this;
    var result = this.simple(object, ['name']);

    result.type = this.typeName(object);
    
    var effects = [];
    object.effects().each(function(i, effect){
        effects.push(this_.serialize(effect));
    });
    if(effects.length > 0)result.effects = effects;

    var children = [];
    object.eachChild(function(child, slot){
        child = this_.serialize(child);
        child.slot = slot;
        children.push(child);
    });
    if(children.length > 0)result.children = children;
    
    return result;
};

/**
 * 指定されたプロパティのみからなるオブジェクトを返す
 */
mabi.Serializer.prototype.simple = function(object, names){
    var result = {};
    $.each(names, function(i, name){
        result[name] = object[name]();
    });
    return result;
};

File: mabi/serializer.py
# -*- coding: utf-8 -*-

import logging

from django.utils import simplejson

from mabi.enchant_class import EnchantClass
from mabi.equipment_class import EquipmentClass
from mabi.element import Element
from mabi.effect import Effect
from mabi.equipment_set import EquipmentSet
from mabi.mob import Mob

logger = logging.getLogger('Serializer')

def deserialize(json):
    '''JSON とモデルの変換を行う

    todo すでに DB に存在する場合は、そのモデルを取得し、プロパティを更新する。
    保存は行わない。
    '''

    json = simplejson.loads(json);
    return deserialize_element(json)

def serialize(model):
    '''オブジェクトを dict と list のみからなる形式へ変換する'''
    return serialize_element(model)


# ----------------------------------------------------------------------

def serialize_element(element):
    '''Effect を dict に変換する'''
    obj = {}
    obj['id'] = unicode(element.key())
    obj['name'] = element.name
    obj['type'] = type(element).__name__

    effects = element.effects
    if effects:
        obj['effects'] = [serialize_effect(e) for e in effects]
    # obj['source'] = element.source

    children = element.children
    if children:
        obj['children'] = [{'slot': e.slot, 'child': serialize_element(e)} for e in children]
    return obj

def serialize_effect(e):
    '''Effect を dict に変換する'''

    b = {
        'op': e.op,
        'param': e.param,
        'min': e.min,
        'max': e.max
        }

    if e.condition:
        b['condition'] = e.condition
    return b

def deserialize_element(json):
    '''JSON(のPython オブジェクトへ変換した物)を Element として Deserialize する'''

    cls = type_to_class(json['type'])

    effects = json.get('effects')
    if effects:
        effects = deserialize_effects(effects)
    else:
        effects = []
    element = cls(name=json['name'], effects=effects)

    children = json.get('children')
    if children:
        for child in children:
            slot = child.get('slot')
            child = child.get('child')
            child = deserialize_element(child)
            child.slot = slot
            element.children.append(child)
    return element

    
def deserialize_effects(json):
    '''effects を JSON から Effect[] へ変換します'''

    result = []
    for effect in json:
        a = Effect(
            op=effect['op'], 
            param=effect['param'], 
            min=effect['min'],
            max=effect.get('max'),
            condition=effect.get('condition'),
            )
        result.append(a)
    return result

def type_to_class(name):
    '''type 名から class を取得する'''
    map = {'Element': Element,
           'EquipmentSet': EquipmentSet,
           'Mob': Mob,
           }
    return map[name]